File: leetcode/00402_remove_k_digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        s = deque()
        for n in num:
            while s and k and n < s[-1]:
                s.pop()
                k -= 1
            s.append(n)
        while k:
            s.pop()
            k -= 1
        return ''.join(s).lstrip('0') or '0'

File: leetcode/00402_remove_k_digits.rs
impl Solution {
    pub fn remove_kdigits(num: String, mut k: i32) -> String {
        let mut ans = String::new();
        for c in num.chars() {
            while k > 0 && !ans.is_empty() {
                let p = ans.pop().unwrap();
                if p > c { k -= 1; }
                else { ans.push(p); break; }
            }
            ans.push(c);
        }
        ans.truncate(ans.len() - k as usize);
        ans = ans.trim_start_matches('0').to_string();
        if ans.len() > 0 { ans } else { "0".to_string() }
    }
}

